File: src/backend/search/brute_force.py
import numpy as np 

class BruteForceNNS: 
    def __init__(self): 
        self.points = None

    def build(self, points): 
        self.points = np.array([x.value for x in points])

    def nearest(self, target, k): 
        query_points = self.points
        target = np.array(target.value)
    
        dist = np.linalg.norm(query_points - target, axis=1)
        sorted_indices = np.argsort(dist)
        indices = sorted_indices[:k]
        distances = dist[sorted_indices[:k]]
        
        return indices, distances

    def farthest(self, target, k):
        query_points = self.points
        target = np.array(target.value)

        dist = np.linalg.norm(query_points - target, axis=1)
        sorted_indices = np.argsort(dist)[::-1]
        indices = sorted_indices[:k]
        distances = dist[sorted_indices[:k]]
    
        return indices, distances
 

   

File: src/frontend/search/brute_force.js
import * as operations from "ivfflat/src/frontend/utils/operations.js"
import * as measures from "ivfflat/src/frontend/utils/measures.js"

export class BruteForceNNS
{
    /** 
     * @param {Object} options 
     */
    constructor() {
        this.points = null
        this.measureFn = measures.euclideanDistance
    }

    // --- TREE CONSTRUCTION ---  // 
    build(points) {
        this.points = points 
    }

    // --- TREE QUERIES --- // 

    nearest(target, k) {
        const results = operations.nearestK(this.points, target, k, this.measureFn)
        const indices = results.map(x => x[0])
        const distances = results.map(x => x[1])
        return [indices, distances]
    }

    farthest(target, k) {
        const results = operations.farthestK(this.points, target, k, this.measureFn)
        const indices = results.map(x => x[0])
        const distances = results.map(x => x[1])
        return [indices, distances]
    }
}